inital implementaion with TODOs to solve

LiorLieberman opened the pull request:
This is just an initial comment of the PR. I did not run `make generate` purposely as I want to get feedback on all the TODOs I listed in the code.

Essentially it comes down to these questions:

1.  Should we mention in the API that `supportedFeatures` should be asc sorted?
2. I want couple of eyes to looks at the annotations I added for the `supportedFeatures` field
3. We are changing `SupportHTTPResponseHeaderModification` feature name to `SupportHTTPRouteResponseHeaderModification` as per the [new standard convention](https://gateway-api.sigs.k8s.io/geps/gep-2162/#standardize-features-and-conformance-tests-names). It will break a few repos who use this constant, I added one for backward compatibility but I am actually against it. I think it is better to make them fix it now before we go to GA.
4. I think experimental features should be also per resource, same as we have for extneded. And not [one for all](https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/suite/features.go#L178C1-L179C1). This will be aligned as well with the [standardized naming and formatting conventions](https://gateway-api.sigs.k8s.io/geps/gep-2162/#standardize-features-and-conformance-tests-names) we established.
5. I am a bit confused about the current requirement in GatewayClass status. Essentially we are saying `Implementations MUST populate status on all GatewayClass resources which specify their controller name` but the conditions field is optional. Can you clarify?

/cc @youngnick @robscott @shaneutt 

/kind feature

**What this PR does / why we need it**:
This PR implements #2163. We are adding new experimental `supportedFeatures` field in GatewayClass status.
**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #2163 

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
Add new experimental `supportedFeatures` field in GatewayClass status
```

LiorLieberman commented:
/hold

robscott commented on conformance/utils/suite/features.go:
```diff
@@ -99,7 +99,11 @@ const (
-	SupportHTTPResponseHeaderModification SupportedFeature = "HTTPResponseHeaderModification"
+	SupportHTTPRouteResponseHeaderModification SupportedFeature = "HTTPRouteResponseHeaderModification"
+
+	//TODO(liorlieberman) I think it is better not to have it and make everyone aligned before GA.
+	// This option is the same as above option but left just for backward compatibility
+	SupportHTTPResponseHeaderModification SupportedFeature = "HTTPRouteResponseHeaderModification"
```
I don't really understand this comment/TODO

robscott commented on conformance/utils/suite/features.go:
```diff
@@ -169,6 +173,7 @@ var MeshCoreFeatures = sets.New(
 // Features - Experimental
 // -----------------------------------------------------------------------------
 
+// TODO(liorlieberman) get consensus about making experimental features per resource.
```
I think I'd rather just tie this specific feature to Gateway since I think all test coverage we have for this also requires Gateway support.

LiorLieberman replied:
tying this to an HTTPRoute as per our offline discussion and as we only have conformance test that checks the HTTPRoute functionality with Destination Port Matching.

robscott reviewed (commented):
Thanks @LiorLieberman!

LiorLieberman commented:
Thanks @robscott.

All the TODOs are also captured in the PR description. The wording in the TODOs in the code are just for me to remember. I will delete them all or convert it to actionable todos after i get feedback on the PRs description.

youngnick commented:
> * Should we mention in the API that `supportedFeatures` should be asc sorted?

Yes.

> * I want couple of eyes to looks at the annotations I added for the `supportedFeatures` field

👀 

> * We are changing `SupportHTTPResponseHeaderModification` feature name to `SupportHTTPRouteResponseHeaderModification` as per the [new standard convention](https://gateway-api.sigs.k8s.io/geps/gep-2162/#standardize-features-and-conformance-tests-names). It will break a few repos who use this constant, I added one for backward compatibility but I am actually against it. I think it is better to make them fix it now before we go to GA.

Yes, I agree that this is okay to break now, so that folks will need to change as part of the GA migration.

> * I think experimental features should be also per resource, same as we have for extneded. And not [one for all](https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/suite/features.go?rgh-link-date=2023-09-08T09%3A35%3A17Z#L178C1-L179C1). This will be aligned as well with the [standardized naming and formatting conventions](https://gateway-api.sigs.k8s.io/geps/gep-2162/#standardize-features-and-conformance-tests-names) we established.

I think this makes sense, but we may need some more discussion here. Let's talk about it in the meeting this week.

> * I am a bit confused about the current requirement in GatewayClass status. Essentially we are saying `Implementations MUST populate status on all GatewayClass resources which specify their controller name` but the conditions field is optional. Can you clarify?

The field is optional so that when you are creating a GatewayClass you don't need to supply a `status` stanza in the YAML.

But implementations must:
- watch for GatewayClasses with their `controllerName` set
- Update the `status.conditions` field with the `Accepted` condition set to `status: true` for all GatewayClasses they accept (which may be only one), and with the `Accepted` condition set to `status: false `for` for all GatewayClasses they do not accept.

Essentially `+optional` and `omitempty` just specify things about the _input_ and _output` of the rendered YAML respectively, not anything about the field itself.

LiorLieberman commented on apis/v1beta1/gatewayclass_types.go:
```diff
@@ -57,6 +57,7 @@ type GatewayClass struct {
 
 	// Status defines the current state of GatewayClass.
 	//
+	// TODO(liorlieberman) must populate status but conditions are optional?
```
resolving per our offline discussion and per's Nick answer

LiorLieberman commented:
Update: 
- Converted features.go to experimental per route
- Converted `RouteDestentaionPortMatching` to `HTTPRouteDestinationPortMatching`
- addressed feedback from comments
- added some typo fixes


PTAL @robscott @youngnick

LiorLieberman commented:
/cc @youngnick @robscott

robscott commented on apis/v1beta1/gatewayclass_types.go:
```diff
@@ -200,6 +200,15 @@ type GatewayClassStatus struct {
 	// +kubebuilder:default={{type: "Accepted", status: "Unknown", message: "Waiting for controller", reason: "Pending", lastTransitionTime: "1970-01-01T00:00:00Z"}}
 	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// TODO(LiorLieberman) explore adding CEL validation to ensure the list is sorted.
```
Mind creating an issue for this instead of a TODO?

robscott reviewed (commented):
Thanks @LiorLieberman! Only one nit, once tests pass this should be good to go.

youngnick commented:
LGTM once that issue is created.

/approve

robscott commented:
Closing in favor of #2461.
